skimage/feature/fisher_vector.py

```diff
@@ -0,0 +1,223 @@
+"""
+fisher_vector.py - Implementation of the Fisher vector encoding algorithm
+
+This module contains the source code for Fisher vector computation. The
+computation is separated into two distinct steps, which are called separately
+by the user, namely:
+
+learn_gmm: Used to estimate the GMM for all vectors/descriptors computed for
+           all examples in the dataset (e.g. estimated using all the SIFT
+           vectors computed for all images in the dataset, or at least a subset
+           of this).
+
+compute_fisher_vector: Used to compute the Fisher vector representation for a
+                       single set of descriptors/vector (e.g. the SIFT
+                       descriptors for a single image in your dataset, or
+                       perhaps a test image).
+
+Reference: Perronnin, F. and Dance, C. Fisher kernels on Visual Vocabularies
+           for Image Categorization, IEEE Conference on Computer Vision and
+           Pattern Recognition, 2007
+
+Origin Author: Dan Oneata (Author of the original implementation for the Fisher
+vector computation using scikit-learn and NumPy. Subsequently ported to
+scikit-image (here) by other authors.)
+"""
+import numpy as np
+
+
+class FisherVectorException(Exception):
+    pass
+
+
+class DescriptorException(FisherVectorException):
+    pass
+
+
+def learn_gmm(descriptors, num_modes, **kwargs):
+    """Function to estimate a GMM given a set of descriptors and number of
+    modes (i.e. Gaussians). This function is essentially a wrapper around
+    the scikit-learn implementation of GMM, namely the
+    sklearn.mixture.GaussianMixture class.
+
+    Due to the nature of the Fisher vector, the only enforce parameter of the
+    underlying scikit-learn class is the covariance_type, which must be 'diag'.
+
+    Parameters
+    -----------
+    descriptors : np.ndarray or list
+        List of NumPy arrays, or a single NumPy array, of the descriptors
+        to use to estimate the GMM. Reason a list of NumPy arrays is
+        permissible is because often when using a Fisher vector encoding,
+        descriptors/vectors are computed separately for each sample/image in
+        the dataset, such as SIFT vectors for each image.
+    num_modes : int
```

Comment:
Does this always need to be specified, or is there also a way to make a reasonable guess?

Reply:
We could set a reasonable default of 32 or 64, as these are common values. However, if we wanted to make a reasonable guess, we could perhaps run a small grid search over this value and choose the GMM with maximum log prob. However, I'm not sure this would be the best approach, and I'm unaware of any other heuristics for this. Let me know your thoughts

Reply:
Yeah, maybe that would be better suited as an example.  Not critical.

Reply:
Although, just adding that idea into the docstring may be helpful already; it gave me an idea of how I would start.

Reply:
I'll add this into the docstring.

Reply:
Done